From 645d413d567fcad9b60749dd109c81b7f86988b9 Mon Sep 17 00:00:00 2001
From: Nikolay Aleksandrov <nikolay@redhat.com>
Date: Fri, 5 Sep 2014 11:03:57 -0400
Subject: [ethernet] sfc: Don't receive packets when the napi budget == 0

Message-id: <1409915048-26438-22-git-send-email-nikolay@redhat.com>
Patchwork-id: 91426
O-Subject: [RHEL7.1 PATCH BZ 1110888 21/32] sfc: Don't receive packets when the napi budget == 0
Bugzilla: 1110888
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: "Eric W. Biederman" <ebiederm@xmission.com>

Processing any incoming packets with a with a napi budget of 0
is incorrect driver behavior.

This matters as netpoll will shortly call drivers with a budget of 0
to avoid receive packet processing happening in hard irq context.

Signed-off-by: "Eric W. Biederman" <ebiederm@xmission.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 75363a4676cdb046242d06dca6e8a9c0a20d6c4a)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/sfc/ef10.c b/drivers/net/ethernet/sfc/ef10.c
index 3b39798..1425663 100644
--- a/drivers/net/ethernet/sfc/ef10.c
+++ b/drivers/net/ethernet/sfc/ef10.c
@@ -1955,6 +1955,9 @@ static int efx_ef10_ev_process(struct efx_channel *channel, int quota)
  int tx_descs = 0;
  int spent = 0;
 
+ if (quota <= 0)
+  return spent;
+
  read_ptr = channel->eventq_read_ptr;
 
  for (;;) {
diff --git a/drivers/net/ethernet/sfc/farch.c b/drivers/net/ethernet/sfc/farch.c
index aa1b169..a087613 100644
--- a/drivers/net/ethernet/sfc/farch.c
+++ b/drivers/net/ethernet/sfc/farch.c
@@ -1248,6 +1248,9 @@ int efx_farch_ev_process(struct efx_channel *channel, int budget)
  int tx_packets = 0;
  int spent = 0;
 
+ if (budget <= 0)
+  return spent;
+
  read_ptr = channel->eventq_read_ptr;
 
  for (;;) {
-- 
1.7.1